vowpalwabbit/example.h

```diff
@@ -55,31 +56,41 @@ typedef union
 IGNORE_DEPRECATED_USAGE_START
 struct example : public example_predict  // core example datatype.
 {
+  example();
+  ~example();
+
+  /// Example contains unions for label and prediction. These do not get cleaned
+  /// up by the constructor because the type is not known at that time. To
+  /// ensure correct cleanup delete_unions must be explicitly called.
+  void delete_unions(void (*delete_label)(void*), void (*delete_prediction)(void*));
+
   // input fields
   polylabel l;
 
   // output prediction
   polyprediction pred;
 
-  float weight;       // a relative importance weight for the example, default = 1
-  v_array<char> tag;  // An identifier for the example.
-  size_t example_counter;
+  float weight = 1.f;  // a relative importance weight for the example, default = 1
```

Comment:
Are all these default values sane? Do we ever have to make a distinction between "value doesn't exist" vs "value isn't set" vs "value is set to default"?

Reply:
These are the values that are used in `setup_example` right now which is effectively like the constructor